### PR TITLE
Add context.attach method to embed files in JSON reports

### LINKED ISSUE #791: How to write a image to behave report?
How to write a image to behave report
How to write a text to behave report

### PR DESCRIPTION
**Pull request by Korijn**
See the test for example.

I attempted to add it to the docs but I'm not sure if my changes are appropriate.

Closes #791

(Travis CI passes: https://travis-ci.org/github/behave/behave/builds/667662398)

### CODE REVIEW

File: features/formatter.json.feature [hunk 1]
```diff
@@ -309,6 +309,48 @@ Feature: JSON Formatter
       But note that "both matched arguments.values are provided as string"
 
 
+    Scenario: Use JSON formatter with feature and one scenario with steps
```
Comment (Korijn) [inline features/formatter.json.feature]: Forgot to come up with a new name for the scenario. Will fix

**Comment by Korijn** [comment]
Ready to merge!

**Comment by Korijn** [comment]
Bump

**Comment by andnasl** [comment]
I really hope this gets merged to master as it would be immensely valuable to add a picture of failures and not force me to use reporting with Allure and their method of attaching screenshot. Tested this with _context.attach("image/png", context.driver.get_screenshot_as_png())_ and got pleasing results with behave2cucumber and cucumber-html-reporter.

**Comment by Korijn** [comment]
I just monkey patched the changes into the library..

**Comment by andnasl** [comment]
That is absolutely amazing! Thank you!

**Comment by Korijn** [comment]
> That is absolutely amazing! Thank you!

You know what they say about monkey patching... with great power comes great responsibility!

**Review by bittner — APPROVED** [review]

File: docs/context_attributes.rst [hunk 2]
```diff
@@ -35,6 +35,7 @@ stdout_capture  scenario  :class:`~StringIO.StringIO`                   If stdou
 stderr_capture  scenario  :class:`~StringIO.StringIO`                   If stderr  capture is enabled.
 table           step      :class:`~behave.model.Table`                  Contains step's table, otherwise None.
 text            step      String                                        Contains step's multi-line text (unicode), otherwise None.
+attach          step      attach(mime_type, data)                       Embed data in formatter output (currently only JSON formatter).
```
Comment (bittner) [inline docs/context_attributes.rst:38]: While it's good to have this documented at least somehow, this information is now added to the list of "Behave Attributes". The `attach` is a member, though, a callable. It somehow doesn't fit.

@jenisys What would be a better way to document this?

File: behave/model.py [hunk 3]
```diff
@@ -1803,6 +1808,7 @@ def run(self, runner, quiet=False, capture=True):
                 #  * Even EMPTY multiline text is available in context.
                 runner.context.text = self.text
                 runner.context.table = self.table
+                runner.context.attach = attach
```
Comment (bittner) [inline behave/model.py:1811]: Is that the right way?

All other predefined `context` attributes are (kind of) read-only members, but this one is a method. It feels like this is trying to escape from a clean image, a clean approach. Or do I get something wrong?

Comment (Korijn) [reply behave/model.py:1811]: Thanks for the review! I can move this method to `runner.py` and add it to the `Context` class. That would be better.

Comment (Korijn) [reply behave/model.py]: I updated the code, what do you think?

**Comment by Korijn** [comment]
Nevermind I messed up with git rebase... give me a minute to fix this..

**Comment by Korijn** [comment]
I addressed your comments and rebased and squashed my changes since the branch was quite old.

File: docs/context_attributes.rst (hunk 2 above)
Comment (Korijn) [reply docs/context_attributes.rst]: I documented the feature in a separate section in the formatters page instead, I think that's more suitable. I also added an example for embedding screenshots with selenium.

**Comment by bittner** [comment]
Thanks! Nice job. :+1: 

I'll discuss this with @jenisys next week. :notebook_with_decorative_cover:

**Comment by jenisys** [comment]
@Korijn You need to rebase your branch with the current HEAD and repush the pull-request.
This fixes the Travis-CI test-run problem with Python2.7.

**Comment by Korijn** [comment]
Done!

**Comment by bittner** [comment]
@jenisys I'm going ahead and merge the changes. The solution looks pretty elegant and provides the basis for other formatters to also support embedding. :chart_with_upwards_trend:

**Comment by bittner** [comment]
Thanks @Korijn! Nice job! :1st_place_medal: 

And thanks you for your patience.

File: docs/formatters.rst [hunk 4]
```diff
@@ -116,3 +116,21 @@ teamcity       :pypi:`behave-teamcity`, a formatter for Jetbrains TeamCity CI te
...
+
+For example:
+
+.. code-block:: python
+
+    @when(u'I open the Google webpage')
+    def step_impl(context):
+        context.browser.get('http://www.google.com')
+        img = context.browser.get_full_page_screenshot_as_png()
+        context.attach("image/png", img)
```
Comment (bittner) [inline docs/formatters.rst:136]: I just figured, we should probably also extend the table of [Behave Attributes](https://behave.readthedocs.io/en/latest/context_attributes.html#behave-attributes) in the "Context Attributes" chapter of our documentation. :thinking:

**Comment by Korijn** [comment]
Were these changes ever released @jenisys ?

**Pull request merged**